### PR TITLE
Zigzag filtration simplex tree

### PR DESCRIPTION
**Pull request by ClementMaria**

### CODE REVIEW

**Review by mglisse — COMMENTED** [review]
Could you suggest an order in which I should look at the various changes? Is there some kind of high-level documentation I could read to understand how things are supposed to work before diving into the details, or some example you recommend reading as a starting point? What does the patch bring, and how are those pieces related?
* link simplices that share the maximal vertex
* build some specific zigzag filtration?
* do some Morse matching?
* ?

What is not included and will be in a separate PR later?
* Actual zigzag persistence?

If you have trouble building the doc, feel free to ask questions, but please look at the generated doc and make sure it contains what a user needs.

Did you check the impact of your changes, for instance on the memory use and speed of a Rips complex? (Do you need help running such tests?) It seems relevant to determine for instance if the python plugin should keep providing just the full-featured version of the simplex tree.

File: src/Simplex_tree/concept/SimplexTreeOptions.h [hunk 1]
```diff
@@ -19,13 +19,19 @@ struct SimplexTreeOptions {
   typedef VertexHandle Vertex_handle;
   /// Must be comparable with operator<.
   typedef FiltrationValue Filtration_value;
-  /// Must be an integer type.
+  /// Must be an integer type. Must be signed if store_morse_pairing == true.
```
Comment (VincentRouvreau) [inline src/Simplex_tree/concept/SimplexTreeOptions.h:22]:
```suggestion
  /// Must be an integer type. Must be signed if store_morse_matching is true.
```

File: src/Simplex_tree/example/CMakeLists.txt [hunk 2]
```diff
@@ -49,3 +49,20 @@ if (TBB_FOUND)
 endif()
 
 add_test(NAME Simplex_tree_example_graph_expansion_with_blocker COMMAND $<TARGET_FILE:Simplex_tree_example_graph_expansion_with_blocker>)
+
+add_executable ( Simplex_tree_rips_zigzag_filtration_from_points rips_zigzag_filtration_from_points.cpp )
+  target_link_libraries(Simplex_tree_rips_zigzag_filtration_from_points Boost::program_options)
+if (TBB_FOUND)
+  target_link_libraries(Simplex_tree_rips_zigzag_filtration_from_points ${TBB_LIBRARIES})
+endif()
+
```
Comment (VincentRouvreau) [inline src/Simplex_tree/example/CMakeLists.txt:58]:
```suggestion
add_test(NAME Simplex_tree_rips_zigzag_filtration_from_points COMMAND $<TARGET_FILE:Simplex_tree_rips_zigzag_filtration_from_points>
      "${CMAKE_SOURCE_DIR}/data/points/tore3D_300.off" -d 3)

```

File: src/Simplex_tree/example/CMakeLists.txt [hunk 3]
```diff
@@ -49,3 +49,20 @@ if (TBB_FOUND)
...
+add_executable ( Simplex_tree_simple_optimized_cofaces_computation simple_optimized_cofaces_computation.cpp )
+if (TBB_FOUND)
+  target_link_libraries(Simplex_tree_simple_optimized_cofaces_computation ${TBB_LIBRARIES})
+endif()
+
```
Comment (VincentRouvreau) [inline src/Simplex_tree/example/CMakeLists.txt:63]:
```suggestion
add_test(NAME Simplex_tree_simple_optimized_cofaces_computation COMMAND $<TARGET_FILE:Simplex_tree_simple_optimized_cofaces_computation>)

```

File: src/Simplex_tree/example/CMakeLists.txt [hunk 4]
```diff
@@ -49,3 +49,20 @@ if (TBB_FOUND)
...
+add_executable ( Simplex_tree_simple_rips_zigzag_filtration simple_rips_zigzag_filtration.cpp )
+if (TBB_FOUND)
+  target_link_libraries(Simplex_tree_simple_rips_zigzag_filtration ${TBB_LIBRARIES})
+endif()
+
```
Comment (VincentRouvreau) [inline src/Simplex_tree/example/CMakeLists.txt:68]:
```suggestion
add_test(NAME Simplex_tree_simple_rips_zigzag_filtration COMMAND $<TARGET_FILE:Simplex_tree_simple_rips_zigzag_filtration>)

```

File: src/Simplex_tree/include/gudhi/Simplex_tree.h [hunk 5]
```diff
@@ -979,8 +1072,17 @@ class Simplex_tree {
...
 
-  Cofaces_simplex_range star_simplex_range(const Simplex_handle simplex) {
-    return cofaces_simplex_range(simplex, 0);
+  Star_simplex_range star_simplex_range(const Simplex_handle simplex) {
+    if constexpr(Options::link_nodes_by_label) {//faster cofaces computation
+      Simplex_vertex_range rg = simplex_vertex_range(simplex);
+      std::vector<Vertex_handle> simp(rg.begin(), rg.end());
+    // must be sorted in decreasing order
+      assert(std::is_sorted(simp.begin(), simp.end(), 
+                            std::greater<Vertex_handle>()));
```
Comment (VincentRouvreau) [inline src/Simplex_tree/include/gudhi/Simplex_tree.h:1081]:
```suggestion
      GUDHI_CHECK(std::is_sorted(simp.begin(), simp.end(), 
                            std::greater<Vertex_handle>()));
```

File: src/Simplex_tree/include/gudhi/Simplex_tree.h [hunk 6]
```diff
@@ -1667,6 +1830,829 @@ class Simplex_tree {
...
+  * deletions of vertices and edges.
+  *
+  * @param[in] zz_edge_fil         range of Zigzag_edge< Simplex_tree >,
+  *                                containing a valid sequence of insertions and
+  *                                removal of vertices and edges.
+  *
+  * @param[in] dim_max             maximal dimension of the dynamic flag complex
+  *                                constructed.
+  *
+  * @param[out]    a range whose iterators are of value_type Simplex_handle.
```
Comment (VincentRouvreau) [inline src/Simplex_tree/include/gudhi/Simplex_tree.h:1932]:
```suggestion
  * @return    a range whose iterators are of value_type Simplex_handle.
```
Doxygen complains about this

File: src/Simplex_tree/include/gudhi/Simplex_tree.h [hunk 7]
```diff
@@ -1667,6 +1830,829 @@ class Simplex_tree {
...
+  * Nodes with label u get affected only if a Node with label v is in their same
+  * siblings set.
+  * We then try to insert "ponctually" v all over the subtree rooted
+  * at Node(u). Each
+  * insertion of a Node with v label induces a local expansion at this
+  * Node (as explained above)
+  * and a sequence of "ponctual" insertion of Node(v) in the subtree
+  * rooted at sibling nodes of the new node, on its left.
+  *
+  * @param[in] u,v              Vertex_handle representing the new edge
```
Comment (VincentRouvreau) [inline src/Simplex_tree/include/gudhi/Simplex_tree.h:2200]:
```suggestion
  * @param[in] u,v              Vertex_handle representing the new edge
  * @param[in] fil              blabla
  * @param[in] dim_max          blabla
```
Doxygen complains about missing `fil` and `dim_max` parameters.

**Comment by VincentRouvreau** [comment]
New PRs are in progress for this new feature

**Pull request closed without merging**